File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int prefixMod = 0, res = 0;
        vector<int> prefixModGrps(k,0);
        prefixModGrps[0]=1;
        for (int i=0; i<nums.size(); i++){
            prefixMod=(prefixMod+nums[i]%k+k)%k;
            res+=prefixModGrps[prefixMod]++;
        }
        return res;
    }
};

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int prefixSum = 0, res = 0;
        int[] modMap = new int[k];
        modMap[0] = 1;
        for(int n: nums){
            prefixSum = (prefixSum + n % k + k) % k; 
            res += modMap[prefixSum]++;
        }
        return res;
    }
}